src/utils/ahash.zig

```diff
@@ -0,0 +1,228 @@
+/// Currently, only the functionality required to replicate the agave Deduper is implemented.
+pub const AHasher = struct {
+    buffer: u64,
+    pad: u64,
+    extra_keys: [2]u64,
+
+    // This constant comes from Kunth's prng (Empirically it works better than those from splitmix32).
+    const MULTIPLE: u64 = 6364136223846793005;
+    const ROTATE: u32 = 23;
+
+    pub fn fromSeed(seed: AHashSeed) AHasher {
+        return .{
+            .buffer = seed.k0,
+            .pad = seed.k1,
+            .extra_keys = .{ seed.k2, seed.k3 },
+        };
+    }
+
+    pub fn hash(self: *AHasher, comptime T: type, data: *const T) void {
+        switch (@typeInfo(T)) {
+            .Int => {
+                self.update(@intCast(data.*));
+            },
+            .Array => |array_info| {
+                self.hashSlice(array_info.child, data);
+            },
+            .Pointer => |pointer| {
+                switch (pointer.size) {
+                    .Slice => self.hashSlice(pointer.child, data.*),
+                    else => {
+                        std.debug.print(
+                            "Unsupported pointer size: type={} pointer.size={}\n",
+                            .{ T, pointer.size },
+                        );
+                        std.debug.panic(
+                            "Unsupported pointer size: type={} pointer.size={}\n",
+                            .{ T, pointer.size },
+                        );
+                    },
+                }
+            },
+            .Struct => |struct_info| {
+                inline for (struct_info.fields) |field| {
+                    self.hash(field.type, &@field(data, field.name));
+                }
+            },
+            .Enum => |_| {
+                self.hash(u32, &@as(u32, @intFromEnum(data.*)));
+            },
+            else => {
+                std.debug.print("Unsupported type: type={} data={any}\n", .{ T, data });
+                std.debug.panic("Unsupported type: type={} data={any}\n", .{ T, data });
+            },
+        }
+    }
+
+    pub fn finish(self: *const AHasher) u64 {
+        const rot: u32 = @intCast(self.buffer & 63);
+        return std.math.rotl(u64, foldedMultiply(self.buffer, self.pad), rot);
+    }
+
+    inline fn hashSlice(self: *AHasher, comptime T: type, data: []const T) void {
+        self.update(data.len);
+        self.write(@as([]const u8, data));
+    }
+
+    inline fn update(self: *AHasher, new_data: u64) void {
+        self.buffer = foldedMultiply(new_data ^ self.buffer, MULTIPLE);
+    }
+
+    inline fn largeUpdate(self: *AHasher, new_data: u128) void {
+        const high: u64 = @intCast(new_data >> 64); // 0..0[0..64]
+        const low: u64 = @intCast((new_data << 64) >> 64); // 0..0[64..127]
+        const combined = foldedMultiply(low ^ self.extra_keys[0], high ^ self.extra_keys[1]);
```

Comment:
what do you think about this? I think it should be faster, not sure if it's a meaningful difference though.
```suggestion
        const first, const second = @as([2]u64, @bitCast(new_data));
        const combined = switch (@import("builtin").cpu.arch.endian()) {
            .little => foldedMultiply(first ^ self.extra_keys[0], second ^ self.extra_keys[1]),
            .big => foldedMultiply(second ^ self.extra_keys[0], first ^ self.extra_keys[1]),
        };
```